pkg/route/controller/ingress/ingress.go

```diff
@@ -0,0 +1,591 @@
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			processNamespaceFn(newObj)
+		},
+	})
+
+	// process a single route
+	processRouteFn := func(obj interface{}) {
+		switch t := obj.(type) {
+		case *routev1.Route:
+			name, ok := hasIngressOwnerRef(t.OwnerReferences)
+			if !ok {
+				return
+			}
+			c.queue.Add(queueKey{namespace: t.Namespace, name: name})
+		default:
+			utilruntime.HandleError(fmt.Errorf("couldn't get key for object %T", obj))
+		}
+	}
+
+	// any change to a route that has the controller relationship to an Ingress
+	routes.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			switch t := obj.(type) {
+			case *routev1.Route:
+				_, ok := hasIngressOwnerRef(t.OwnerReferences)
+				return ok
+			}
+			return true
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    processRouteFn,
+			DeleteFunc: processRouteFn,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				processRouteFn(newObj)
+			},
+		},
+	})
+
+	// process a single ingress
+	processIngressFn := func(obj interface{}) {
+		switch t := obj.(type) {
+		case *extensionsv1beta1.Ingress:
+			c.queue.Add(queueKey{namespace: t.Namespace, name: t.Name})
+		default:
+			utilruntime.HandleError(fmt.Errorf("couldn't get key for object %T", obj))
+		}
+	}
+
+	// changes to ingresses
+	ingresses.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			// switch t := obj.(type) {
+			// case *extensionsv1beta1.Ingress:
+			// 	return true
+			// }
+			return true
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    processIngressFn,
+			DeleteFunc: processIngressFn,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				processIngressFn(newObj)
+			},
+		},
+	})
+
+	c.syncs = []cache.InformerSynced{
+		ingresses.Informer().HasSynced,
+		routes.Informer().HasSynced,
+		secrets.Informer().HasSynced,
+		services.Informer().HasSynced,
+	}
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting controller")
+
+	if !cache.WaitForCacheSync(stopCh, c.syncs...) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down controller")
+}
+
+func (c *Controller) worker() {
+	for c.processNext() {
+	}
+	glog.V(4).Infof("Worker stopped")
+}
+
+func (c *Controller) processNext() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.sync(key.(queueKey))
+	c.handleNamespaceErr(err, key)
+
+	return true
+}
+
+func (c *Controller) handleNamespaceErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
```

Comment:
We don't usually do this.  The last AddRateLimited is order of tens of minutes.  I keeps us from having to worry about permafails